lib/core/monitor_impl.toit

```diff
@@ -44,104 +44,121 @@ class __Monitor__:
+    while true:
+      // Check for task cancelation.
+      if is_non_critical and self.is_canceled_: throw CANCELED_ERROR
+      // Check for task timeout.
+      if deadline and Time.monotonic_us >= deadline: throw DEADLINE_EXCEEDED_ERROR
+      // If the monitor isn't owned by anyone at this point, we are ready
+      // to take it.
+      if not owner_: break
+      // Wait until notified. When we get back the monitor might be owned by
+      // someone else.
+      wait_ self
+
     owner_ = self  // Take lock.
     try:
       block.call
     finally:
       owner_ = null
-      notify_next_
-      notify_awaits_
+      // State may have changed as part of running the locked method.
+      resume_ --state_changed
+      // To guarantee some level of fairness, we yield to avoid letting
+      // the calling task starve the others.
+      if is_non_critical: yield
 
-  // Wait for mutex to be free.
   wait_ self:
-    done := false
-    while not done:
-      // Add self to end of waiters list.
-      tail := waiters_tail_
-      if tail:
-        tail.next_blocked_ = self
-        waiters_tail_ = self
-      else:
-        waiters_head_ = waiters_tail_ = self
-      suspend_ self
-      done = owner_ == null
-    assert: owner_ == null
+    // Add self to end of waiters list.
+    tail := waiters_tail_
+    if tail:
+      tail.next_blocked_ = self
+      waiters_tail_ = self
+    else:
+      waiters_head_ = waiters_tail_ = self
+    suspend_ self
 
-  // Wait for someone to notify.
   await_ self:
-    done := false
-    while not done:
-      // Add self to end of awaiters list.
-      tail := await_tail_
-      if tail:
-        tail.next_blocked_ = self
-        await_tail_ = self
-      else:
-        await_head_ = await_tail_ = self
-      suspend_ self
-      done = owner_ == null
-    assert: owner_ == null
+    // Add self to end of awaiters list.
+    tail := await_tail_
+    if tail:
+      tail.next_blocked_ = self
+      await_tail_ = self
+    else:
+      await_head_ = await_tail_ = self
+    suspend_ self
 
-  notify_all_:
-    // If there is an owner, ignore. We'll notify all waiting once we unlock.
-    if owner_: return
-    notify_next_
-    notify_awaits_
-
-  // Unblock next task waiting to run locked methods.
-  notify_next_:
-    assert: owner_ == null
-    waiter := waiters_head_
-    if waiter:
-      // Unlink.
-      waiters_head_ = waiter.next_blocked_
-      waiter.next_blocked_ = null
-      if not waiters_head_: waiters_tail_ = null
-      // Now resume it.
-      waiter.resume_
-
-  // Unblock all tasks waiting for a condition to be fulfilled.
-  notify_awaits_:
+  resume_ --state_changed/bool:
     assert: owner_ == null
+    // If the state changed, we first resume the tasks waiting to
+    // re-evaluate their conditions. These are tasks that have
+    // already acquired the lock in the past, so it makes sense to
+    // let them run first.
     waiter := await_head_
```

Comment:
Nit: it's odd to me that the waiting queues are named "waiters" and "await" instead of "waiters" and "awaiters".